index.bs

```diff
@@ -1370,77 +880,818 @@ What configuration is the device in after it resets?
+    Promise<undefined> selectAlternateInterface(octet interfaceNumber, octet alternateSetting);
+    Promise<USBInTransferResult> controlTransferIn(USBControlTransferParameters setup, unsigned short length);
+    Promise<USBOutTransferResult> controlTransferOut(USBControlTransferParameters setup, optional BufferSource data);
+    Promise<undefined> clearHalt(USBDirection direction, octet endpointNumber);
+    Promise<USBInTransferResult> transferIn(octet endpointNumber, unsigned long length);
+    Promise<USBOutTransferResult> transferOut(octet endpointNumber, BufferSource data);
+    Promise<USBIsochronousInTransferResult> isochronousTransferIn(octet endpointNumber, sequence<unsigned long> packetLengths);
+    Promise<USBIsochronousOutTransferResult> isochronousTransferOut(octet endpointNumber, BufferSource data, sequence<unsigned long> packetLengths);
+    Promise<undefined> reset();
+  };
 </xmp>
 
-A <dfn>control transfer</dfn> is a special class of USB traffic most commonly
-used for configuring a device. It consists of three stages:
-<a lt="setup stage">setup</a>, <a lt="data stage">data</a> and
-<a lt="status stage">status</a>. In the <dfn>setup stage</dfn> a
-<dfn>setup packet</dfn> is transmitted to the device containing request
-parameters including the transfer direction and size of the data to follow. In
-the <dfn>data stage</dfn> that data is either sent to or received from the
-device. In the <dfn>status stage</dfn> successful handling of the request is
-acknowledged or a failure is signaled.
+Instances of {{USBDevice}} are created with the <a>internal
+slots</a> described in the following table:
+
+<table class="data" dfn-for="USBDevice" dfn-type="attribute">
+  <tr>
+    <th><a>Internal Slot</a></th>
+    <th>Initial Value</th>
+    <th>Description (non-normative)</th>
+  </tr>
+  <tr>
+    <td><dfn>\[[configurations]]</dfn></td>
+    <td>An empty <a>sequence</a> of {{USBConfiguration}}</code></td>
+    <td>
+      All configurations supported by this device.
+    </td>
+  </tr>
+  <tr>
+    <td><dfn>\[[configurationValue]]</dfn></td>
+    <td>&lt;always set in prose></td>
+    <td>
+      The current configuration value of the device.
+    </td>
+  </tr>
+  <tr>
+    <td><dfn>\[[selectedAlternateSetting]]</dfn></td>
+    <td>An <a>empty</a> list of <a>integer</a></td>
+    <td>
+      The current alternate setting for each interface on the current configuration.
+    </td>
+  </tr>
+  <tr>
+    <td><dfn>\[[claimedInterface]]</dfn></td>
+    <td>An <a>empty</a> list of <a>boolean</a></td>
+    <td>
+      The claimed status for each interface on the current configuration.
+    </td>
+  </tr>
+</table>
+
+<div algorithm="To find the device descriptor for the connected usb device">
+    <dfn>To find the device descriptor for the connected usb device</dfn>,
+    performing the following steps:
+    1. Let |deviceDescriptor| be the <a>device descriptor</a> of the connected
+        device by performing <a>Get Descriptor</a> with
```

Comment:
I think these algorithms should state the assumption that these descriptors have already been read from the device as part of originally enumerating the device. This is how Chromium implements it and there's no need to imply that the data would be re-requested for any of these steps.

Reply:
I added an additional note in https://github.com/WICG/webusb/pull/212/commits/1a96f493fc285da99d2b812d809c241388197beb#diff-5e793325cd2bfc452e268a4aa2f02b4024dd9584bd1db3c2595f61f1ecf7b985R2419 to call out the implementation can choose to optimize these non-necessary get descriptor traffic to the device.

Let me know how I think by specifying it in this way.